### PR TITLE
fix: `missing-docs-allow-unused` for `_`-prefixed associated consts

### LINKED ISSUE #17052: `missing-docs-allow-unused` doesn't catch associated `const`s.
### Summary

I suggested `missing-docs-allow-unused` in #14413 which has been useful, but it seems it was a bit constrained in its scope and only applies to fields. However I ran into the case today where the `missing-docs-allow-unused` behaved unexpectedly - it didn't silence an error in a `const _SIZE_TEST: () = { assert!(...) }` declaration where I'd have expected it to.

Could it be expanded to include any declaration that has an identifier that is `_`-prefixed?

### Lint Name

missing_docs_in_private_items

### Reproducer

I tried this code:

```rust
//! Test
pub struct Foo(pub usize);

impl Foo {
    const _SIZE_ASSERT: () = {
        assert!(
            core::mem::size_of::<Self>().is_multiple_of(4),
            "not a multiple of 4"
        );
    };
}
```

```toml
# clippy.toml
missing-docs-allow-unused = true
```

I saw this happen:

```
error: missing documentation for an associated constant
  --> src/lib.rs:5:5
   |
 5 | /     const _SIZE_ASSERT: () = {
 6 | |         assert!(
 7 | |             core::mem::size_of::<Self>().is_multiple_of(4),
 8 | |             "not a multiple of 4"
 9 | |         );
10 | |     };
   | |______^
```

I expected no clippy lint to emit.

### Version

```text
clippy 0.1.90 (1159e78c47 2025-09-14)
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by Kokoro2336**
changelog: [`missing_docs_in_private_items`] add `missing-docs-allow-unused` for `_`-prefixed associated consts

Fixes rust-lang/rust-clippy#17052

### CODE REVIEW

**Comment by samueltardieu** [comment]
r? clippy

File: clippy_lints/src/missing_doc.rs [hunk 1]
```diff
@@ -69,6 +69,10 @@ impl MissingDoc {
         }
     }
 
+    fn is_allow_unused_name(&self, name: &str) -> bool {
+        self.allow_unused && name.starts_with('_')
+    }
```
Comment (ada4a) [inline clippy_lints/src/missing_doc.rs:74]: Currently, this requires eagerly turning `Symbol` into `&str`, which is a bit expensive. It would be better to do
```suggestion
    fn is_allow_unused_name(&self, name: Symbol) -> bool {
        self.allow_unused && name.as_str().starts_with('_')
    }
```
so that the conversion only happens if `self.allow_unused` is `false`.

File: clippy_lints/src/missing_doc.rs [hunk 2]
```diff
@@ -200,6 +205,7 @@ impl<'tcx> LateLintPass<'tcx> for MissingDoc {
             && self.automatically_derived_depth == 0
             && self.in_body.is_none()
             && !item.span.from_expansion()
+            && !self.is_allow_unused_name(item.ident.name.as_str())
```
Comment (ada4a) [inline clippy_lints/src/missing_doc.rs:208]: could you please add tests for this as well, under [`TraitPub`](https://github.com/rust-lang/rust-clippy/blob/2f1f5537b3aa771380a008b1325e11c5ce8cda61/tests/ui-toml/missing_docs_in_private_items/missing_docs_in_private_items.rs#L84), [`TraitCrate`](https://github.com/rust-lang/rust-clippy/blob/2f1f5537b3aa771380a008b1325e11c5ce8cda61/tests/ui-toml/missing_docs_in_private_items/missing_docs_in_private_items.rs#L190), etc.?

File: tests/ui-toml/missing_docs_in_private_items/missing_docs_in_private_items.rs [hunk 3]
```diff
@@ -1165,3 +1165,8 @@ external! {
         pub struct VisFromOutside; //~ missing_docs_in_private_items
     )}
 }
+
+pub struct SizeCheck;
+impl SizeCheck {
+    const _SIZE_ASSERT: () = {}; //~[default,crate_root] missing_docs_in_private_items
+}
```
Comment (ada4a) [inline tests/ui-toml/missing_docs_in_private_items/missing_docs_in_private_items.rs:1172]: instead of adding a separate test, could you please move this assoc const to under `StructFieldPub` (and also duplicate it under `StructFieldCrate` etc.)?

File: clippy_lints/src/missing_doc.rs [hunk 4]
```diff
@@ -166,6 +170,7 @@ impl<'tcx> LateLintPass<'tcx> for MissingDoc {
         };
 
         if !item.span.from_expansion()
+            && !self.is_allow_unused_name(cx.tcx.item_name(item.owner_id.def_id).as_str())
```
Comment (ada4a) [inline clippy_lints/src/missing_doc.rs:173]: could you please add some tests for this?

**Review by ada4a — CHANGES_REQUESTED** [review]
Apologies for the late review...

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17073/ba8b78b0a89a83d227540770a4274c2840a046aa..2f1f5537b3aa771380a008b1325e11c5ce8cda61)*
